### DISCUSSION TITLE
Writing an image produces an octet stream

### DISCUSSION THREAD

**Discussion by ghost** [dlemstra/magick-wasm#62] [Q&A]
Hi. I've found that writing some images behave strangely. The attached image, when written as PNG (well I suppose it's okay to ignore the fact that the source is a PNG already), is identified by wasmagic as application/octet-stream, while if written as JPG, identified correctly as image/jpeg. When written as WEBP, it's also identified as application/octet-stream. How can I deal with this?

![WATERALPHA64](https://user-images.githubusercontent.com/112320206/197350242-df7d1151-309f-4d95-8acb-a38eb54073eb.png)

P.S The image is above. It's invisible, but contains alpha information.

**Comment by dlemstra** [comment]
Can you share a bit of sample code that demonstrates what you are doing? I am having trouble to understand your question,

**Reply by ghost** [reply]
So I have some example code: https://gist.github.com/Vasia-Assembler/a6b0596311e1ef0bea128f8c4124c0ec

It loads an image, writes it as PNG, and then reads it and writes as PNG once again.

When loading most images, the output is, as I expected: `image/png
Uint8Array(536)
Uint8Array(536)`

But rarely, I stumble upon images which behave unexpectedly and refuse to be read after being written as PNG, and running this code with those images is: `application/octet-stream Uint8Array(59486) Uncaught Error: NoDecodeDelegateForThisImageFormat ```' @ error/blob.c/BlobToImage/460 ...`

I haven't run into problems with those images when I was using the other wasm imagemagick library few years ago, so I guess I have to ask here if something could be done about this.

I attached an image which will also cause this strange behaviour.
![TELEPORT_ANIM](https://user-images.githubusercontent.com/112320206/197388130-3e5b7e41-6a44-45ea-b5a5-8c3348ccdeb6.png)

**Comment by dlemstra** [comment]
The 'data' buffer will be gone at some point after the method has returned because that buffer is on a separate buffer that will be freed after the method returns. So you will need to make a copy of that 'data' buffer. You will need to do something like this:

```js
const a = new Promise<Uint8Array>((resolve) => {
  ImageMagick.read(view, (img) => {
    img.autoOrient()
    img.write((data) => {
      const output = new Uint8Array(data);
      resolve(output);
    }, MagickFormat.Png)
  })
})
```
